server/config.js

```diff
@@ -0,0 +1,5 @@
+module.exports = require('nconf')
+    .argv()
+    .env({ lowerCase: true })
+    .file('environment', { file: `config/${process.env.NODE_ENV}.json` })

```

Comment:
I think we should use the __dirname syntax to get the current directory for the executing file, and not rely on the user running the project from the correct directory. While this currently works in this file, it's not really intuitive to the user as to where the config files are located in the project. 

Also, `app.js` needs to be updated to use this module, and despite using the same syntax here for the file location it does not work in `app.js`. The inconsistency concerns me, which is why I think __dirname is a better plan here.

```suggestion
    .file('environment', { file: `${__dirname}/../config/${process.env.NODE_ENV}.json` })
    .file('defaults', { file: `${__dirname}/../config/default.json` });
```

Reply:
Elaborating because I think I figured out why it works right now in `config.js` but not in `app.js`.

Since `app.js` calls `process.chdir(__dirname)` it ends up changing the path for its execution to the `server` sub-folder, in which there is no `config` directory. `config.js` does not do that, and since it's being required in `server.js` the `config` folder does exist. 

I don't think `chdir` is the way to go, I really think we should use `__dirname` in our static paths to non-js files.

Reply:
Yes, this got broken in #1959 when the `config` initialization was left in `server.js` but `process.chdir(__dirname)` was moved out so it could no longer find the config files.

@AlexeySachkov I think @jeddai 's suggested fix is the way to go here.

Reply:
@jeddai, @calculuschild, Oops, sorry folks for breaking this. I was suspicious about this `process.chdir` line, but apparently not enough. I will prepare a fix

Reply:
Ok, I did the change locally, but it doesn't seem to work correctly:

```js
const path = require('path');

const defaultConfigPath = path.resolve(__dirname, '../config/default.json');

console.debug('path.resolve(__dirname, "../config/default.json") in config.js ' + defaultConfigPath);
console.debug('"${__dirname}/../config/default.json" in config.js ' + `${__dirname}/../config/default.json`);

module.exports = require('nconf')
    .argv()
    .env({ lowerCase: true })
    .file('environment', { file: `${__dirname}/../config/${process.env.NODE_ENV}.json` })
    .file('defaults', { file: defaultConfigPath });

const fs = require('fs');

fs.readFile(path.resolve(__dirname, '../config/default.json'), 'utf8', (err, data) => {
    if (err) throw err;
    console.log('default config file content read from config.js: ' + data);
  });
```

I'm launching the server out of `server/` sub-directory and this is what I see (`server.js` was explicitly tweaked to use port `8080` as a fallback):
```
$ node ../server.js 
path.resolve(__dirname, "../config/default.json") in config.js C:\Users\asachkov\github.com\naturalcrit\homebrewery\config\default.json
"${__dirname}/../config/default.json" in config.js C:\Users\asachkov\github.com\naturalcrit\homebrewery\server/../config/default.json
config.get(web_port) in server.js: undefined
server on port: 8080
default config file content read from config.js: {
        "host" : "homebrewery.local.naturalcrit.com:8000",
        "naturalcrit_url" : "local.naturalcrit.com:8010",
        "secret" : "secret",
        "web_port" : 8000,
        "enable_v3" : true
}
```

The path produced by `path.resolve` is correct (path which uses `${__dirname}` uses different separators so I wasn't sure about it, but it doesn't work as well), the file is definitely accessible, but `config` object doesn't seem to be able to grab the data out of it? What am I missing here?

Reply:
Ah you're on Windows. I think `path.resolve` is going to be for sure required, in hindsight. Seems like it does some OS normalization and switches the `/` to `\` for Windows (I'm on Windows too, I just use it from within WSL so I didn't notice that).

I think this should do it:
```js
const path = require('path');

module.exports = require('nconf')
    .argv()
    .env({ lowerCase: true })
    .file('environment', { file: path.resolve(__dirname, `../config/${process.env.NODE_ENV}.json`) })
    .file('defaults', { file: path.resolve(__dirname, '../config/default.json') });
```

Reply:
As for the port issue, I tried to replicate it using the file as described in the comment above and was unsuccessful. Does what I did in this image generally seem like the same thing you did?

![image](https://user-images.githubusercontent.com/4909307/151427875-7ad9a271-96c3-4559-8226-ef79fc0c1949.png)

Reply:
I've just returned `config.js` to use plain relative paths and started a server from a root project directory, where everything is supposed to be working and it still doesn't.

I've also spend several minutes in debugger going through `nconf` code: it was definitely able to read the file correctly, but it by some reason can't get any values out of it and I have no idea why.

I will now try googling for the issue, but my fallback plan is to simply add `path.resolve` and hope that everything will work on your side, folks.

Reply:
I'm starting to wonder if it's a Windows thing... Which quite frankly would suck.

Reply:
This could be some local issue, probably I'm running different version of node (`v16.13.1`) and I've encountered some incompatibility with packages we use?

Right now, I'm loading `defaults.json` the way it were loaded, using relative path without `__dirname` or `path.resovle`; I'm launching the server from the root directory; I've saved `nconf` object into a local variable before storing it to `module.exports` in order to use it in `config.json` and this is what I get:

```
$ node server.js
config.js: config.get(web_port) 8000
server.js: config.get(web_port) undefined
server on port: 8080
```

Values loaded by config magically disappear when config object is accessed by `server.js`. I have no idea why

Reply:
Does the nconfig object actually store the config values when it's initiated or does it just store like... a path to the config values? So when server.js tries to do config.get those paths are off by one folder?

Reply:
I've console logged it out before and it definitely seems like it's a path reference and not actual values.

Reply:
Ok, so if we restore `process.chdir(__dirname)` to `server.js` so the "root" folder for everything is based on that, can we go back to `.file('environment', { file: \`config/${process.env.NODE_ENV}.json\` })` as an absolute path?

Reply:
Seems we can configure in-memory storage of the values https://www.npmjs.com/package/nconf#Storage-Engines:

> Memory
> A simple in-memory storage engine that stores a nested JSON representation of the configuration. To use this engine, just call .use() with the appropriate arguments. All calls to .get(), .set(), .clear(), .reset() methods are synchronous since we are only dealing with an in-memory object.
> ```
>nconf.use('memory');
> ```

Reply:
If that does what it sounds like, it might work.

I think we should still restore `processs.chdir(__dirname)` to `server.js` so the root directory of the app is consistent when it's needed.

Reply:
> I've console logged it out before and it definitely seems like it's a path reference and not actual values

Hmm, I went on a quick 20 minute adventure in an out with a debugger and I definitely saw it reading a file within `.file()` call.

Adding `use('memory')` doesn't help me regardless of whether I add it before or after calling `.file()`, but what helps me is changing first `.file()` argument from `defaults` to `defaults2` - it seems like there are special cases for configurations named `defaults` and `overrides`, which effectively drop our values

Reply:
I think the issue there is that regardless of using `.use('memory')` it's still not a singleton, so each time it gets required by something it's going to be a new instance, created then.

Reply:
I think the following issue could be relevant: indexzero/nconf#142

Reply:
> I think the following issue could be relevant: indexzero/nconf#142

Nope, `grep` doesn't show anyone requiring `nconf` or importing from `nconf`

Reply:
Have you tried with the following nconf setup:
```js
const path = require('path');

module.exports = require('nconf')
    .argv()
    .env({ lowerCase: true })
    .file('environment', { file: path.resolve(__dirname, `../config/${process.env.NODE_ENV}.json`) })
    .file('defaults', { file: path.resolve(__dirname, '../config/default.json') });
```

Reply:
> Have you tried with the following nconf setup:

Yes, and it doesn't work.

My current guess is that I'm encountering [module caching caveat](https://nodejs.org/api/modules.html#module-caching-caveats) with some of our modules being re-resolved and therefore some of loaded values being overwritten/dropped. At least that is my current guess

Reply:
Ok, folks, I've pushed an update: 4e0ab4b39367cfba4361d57e7fea23e66ee816ef, in which I've applied `path.resolve` approach to load `.json` files as suggested by @jeddai. Also I've changed `defaults` to `default` to fix the problem I was experiencing.

With these changes I'm able to properly launch server from both root directory and from `server/` directory - it would be great to have confirmation that it works on your side as well, @jeddai, @calculuschild.

Reply:
Hmmm that seems most likely. @calculuschild does the branch load the config correctly for you now?